Add schema.org video metadata.

Bopchy opened the pull request:
Resolves #578 

**Overall change:** _Adds video metadata to schema.org._
 
**Code changes:**

- _Add metadata helmet tag to Video component._
- _Add metadata propType to Video component._
- _Currently, the duration and uploadDate are causing the below errors on the structured data testing tool._ 
<img width="576" alt="Screenshot 2019-05-10 at 18 13 31" src="https://user-images.githubusercontent.com/12303856/57540703-fa48cc00-7355-11e9-9718-f70a174a0a2f.png">

- _UploadDate will be added by #1685._
---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

Commit:
Add videoMetadata to Video container.
- Add metadata helmet tag to Video component.
- Add metadata propType to Video component.

Commit:
Remove comments. Fix some eslint issues.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

Commit:
Remove Helmet script from Video component.

Commit:
Add tests.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

Commit:
Merge branch 'latest' into 578-add-schema.org-video-metadata

Commit:
Fix eslint issues.

Commit:
Merge branch '578-add-schema.org-video-metadata' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

sareh commented on src/app/containers/Video/videoMetadata.js:
```diff
@@ -0,0 +1,31 @@
+      },
+    };
+  }
+  return undefined;
```
Instead of returning undefined at the end, it'd be good to have an early return at the start. 
e.g. `if(!data || data.length < 1) { return null; }`

Bopchy replied:
I thought it might be better to return `undefined` instead of `null` so that when `!data || data.length < 1` nothing will be returned by the script tag in index.js

sareh commented on src/app/containers/Video/videoMetadata.test.js:
```diff
@@ -0,0 +1,89 @@
+
+  it('handles aresMediaMetadata type not being present', () => {
+    const metadata = videoMetadata(noAresMediaMetadata);
+    const output = { video: { '@list': [] } };
```
If there is no `aresMediaMetadata`, we should not return any JSON at all.

sareh commented on src/app/containers/Video/videoMetadata.test.js:
```diff
@@ -0,0 +1,89 @@
+    const output = {
+      video: {
+        '@list': [
+          {
```
It'd be good to also add a unit test for multiple videos - to show we output the metadata within this `@list` array

sareh commented on src/app/containers/Video/index.jsx:
```diff
@@ -17,10 +19,21 @@ const VideoContainer = ({ blocks }) => {
-      <Video />
-      {captionBlock ? <Caption block={captionBlock} video /> : null}
-    </Figure>
+    (
```
Instead of returning with parentheses, you can return using fragments 
```
return (
  <>
    <Helmet>
      ...
    </Helmet>
    <Figure>
    ...
    </Figure>
  </>
);
```
https://reactjs.org/docs/fragments.html#short-syntax

sareh commented on src/app/containers/Video/videoMetadata.test.js:
```diff
@@ -0,0 +1,89 @@
+import videoMetadata from './videoMetadata';
+import { aresMediaBlock } from './fixtureData';
+
+const noAresMediaMetadata = {
```
You should move this into the `./fixtureData` file, where the other aresMedia block definitions are.

dr3 commented:
👋 Apologies, this PR needs a bump and changelog update :) Details can be found [here](https://github.com/bbc/simorgh/blob/a75ad1fbf4a720bf6147d7f57ee3d8e3a7337632/CONTRIBUTING.md#versioning-and-changelog)

sareh commented on src/app/containers/Video/videoMetadata.js:
```diff
@@ -0,0 +1,31 @@
+const videoObject = 'VideoObject';
+
+const videoMetadata = blocks => {
+  const data = deepGet(['model', 'blocks'], blocks);
```
Would be good to name this as `aresMediaBlocks` rather than `data` for clarity.

sareh reviewed (changes requested):
Looks great so far. I've added some comments.

Commit:
Implement requested code review changes.
- Move test aresMetadata data to fixtureData.js.
- Add unit test for multiple aresMetadata types.
- Implement early return for data validity checks.
- Replace parentheses with fragments in Video component.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

Commit:
Update CHANGELOG.md. Update package.json.

Commit:
Remove quotes from fixture data.

sareh commented on src/app/containers/Video/videoMetadata.js:
```diff
@@ -0,0 +1,32 @@
+  };
+
+  if (!aresMediaBlocks || aresMediaBlocks.length < 1) {
+    return undefined;
```
It'd be better to return `null` than undefined.

sareh commented on src/app/containers/Video/videoMetadata.js:
```diff
@@ -0,0 +1,32 @@
+  if (!aresMediaBlocks || aresMediaBlocks.length < 1) {
+    return undefined;
+  }
+  aresMediaBlocks.forEach(block => {
```
You can use a filter function here instead, so you don't need to nest a `forEach` with an `if`.

sareh commented on src/app/containers/Video/videoMetadata.js:
```diff
@@ -0,0 +1,32 @@
+      thumbnailUrl: `https://${deepGet(['model', 'imageUrl'], block)}`,
+    });
+  });
+  return listContent.length > 0 ? metadata : undefined;
```
Likewise here, returning `null` would be preferable rather than `undefined`.

sareh commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "simorgh",
-  "version": "1.0.0",
+  "version": "1.1.0",
```
Apologies for the confusion regarding changelogs and version number updates. It's now been clarified that we won't have a Changelog file and won't increment the Simorgh package version in PRs.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

Commit:
Use filter function to remove nested if. Use null in favor of undefined.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 578-add-schema.org-video-metadata

Commit:
Return null vs script if metadata is not defined.

sareh commented on src/app/containers/Video/index.jsx:
```diff
@@ -12,15 +14,23 @@ import { filterForBlockType } from '../../helpers/blockHandlers';
+    <>
+      {metadata ? (
+        <Helmet>
+          {<script type="application/ld+json">{metadata}</script>}
```
The contents of this script tag need to be JSON.stringified, since Helmet expects that.
The way to check the output is working is by running Storybook locally and going to the [Video Container story](http://localhost:9001/?selectedKind=Video%20Container&selectedStory=with%20a%20caption) 
This is the error that is currently visible: 
<img width="1402" alt="Screen Shot 2019-05-13 at 10 49 51" src="https://user-images.githubusercontent.com/3028997/57612537-ecb95f00-756c-11e9-8e1e-9b5b9c7d0906.png">

Bopchy replied:
Sorry about that. Fixed.

Commit:
JSON.stringify metadata.

Commit:
Merge branch 'latest' into 578-add-schema.org-video-metadata

sareh reviewed (dismissed):
👍 Looks good, locally. I see the relevant JSON in a script tag in the head of the Video container iframes in Storybook. 
<img width="600" alt="Screen Shot video container iframe" src="https://user-images.githubusercontent.com/3028997/57681463-ab39ba00-7627-11e9-9b5a-5126a745d512.png">

http://localhost:9001/?selectedKind=Video%20Container

twinlensreflex commented on src/app/containers/Video/index.jsx:
```diff
@@ -12,15 +14,27 @@ import { filterForBlockType } from '../../helpers/blockHandlers';
+      {metadata ? (
+        <Helmet>
+          {
+            <script type="application/ld+json">
```
```suggestion
            <script type="application/json+ld">
```
@Bopchy should it be `json+ld` or `ld+json`.. does the distinction matter? I don't know but it's just something I noticed while reviewing..

sareh replied:
It should be `application/ld+json`. It's correct in the PR. https://developers.google.com/search/docs/guides/intro-structured-data

twinlensreflex replied:
sweet! thanks 😄

Commit:
Merge branch 'latest' into 578-add-schema.org-video-metadata

Commit:
Merge branch 'latest' into 578-add-schema.org-video-metadata

oluoluoxenfree reviewed (dismissed):
LGTM :)

jamesbrumpton committed:
Merge branch 'latest' into 578-add-schema.org-video-metadata

sareh committed:
Merge branch 'latest' into 578-add-schema.org-video-metadata

sareh committed:
Remove duplication introduced by merging in latest

oluoluoxenfree reviewed (approved):
LGTM :)

jamesbrumpton commented:
LGTM. Happy for this to be merged.
